File: src/tools/pmd-cpd/eval-repos/synthetic/csharp/CrossFileA.cs
/// <summary>
/// C# file A for cross-file duplication detection.
/// </summary>

using System;
using System.Collections.Generic;

namespace Synthetic.CrossFile
{
    public class OrderItem
    {
        public decimal Price { get; set; }
        public int Quantity { get; set; } = 1;
        public decimal Discount { get; set; }
        public string Name { get; set; }
    }

    public class Order
    {
        public string Id { get; set; }
        public string CustomerName { get; set; }
        public string Date { get; set; }
        public List<OrderItem> Items { get; set; }
        public decimal Subtotal { get; set; }
        public decimal Shipping { get; set; }
        public decimal Tax { get; set; }
        public decimal Total { get; set; }
    }

    public static class OrderCalculator
    {
        public static decimal CalculateOrderTotal(List<OrderItem> items)
        {
            decimal total = 0.0m;
            foreach (var item in items)
            {
                decimal price = item.Price;
                int quantity = item.Quantity;
                decimal discount = item.Discount;
                decimal itemTotal = price * quantity * (1 - discount / 100);
                total += itemTotal;
            }
            return Math.Round(total, 2);
        }

        private static readonly Dictionary<string, decimal> ShippingRates = new()
        {
            { "US", 5.99m },
            { "CA", 8.99m },
            { "UK", 12.99m },
            { "DE", 14.99m },
            { "FR", 14.99m },
            { "AU", 19.99m }
        };

        public static decimal ApplyShippingCost(decimal subtotal, string country)
        {
            decimal baseRate = ShippingRates.GetValueOrDefault(country, 24.99m);
            if (subtotal > 100) return subtotal;
            return subtotal + baseRate;
        }

        private static readonly Dictionary<string, decimal> TaxRates = new()
        {
            { "CA", 0.0725m },
            { "NY", 0.08m },
            { "TX", 0.0625m },
            { "FL", 0.06m },
            { "WA", 0.065m }
        };

        public static decimal ApplyTax(decimal subtotal, string state)
        {
            decimal rate = TaxRates.GetValueOrDefault(state, 0.0m);
            decimal tax = subtotal * rate;
            return Math.Round(subtotal + tax, 2);
        }

        public static string FormatOrderSummary(Order order)
        {
            var lines = new List<string>();
            lines.Add(new string('=', 50));
            lines.Add("ORDER SUMMARY");
            lines.Add(new string('=', 50));
            lines.Add($"Order ID: {order.Id ?? "N/A"}");
            lines.Add($"Customer: {order.CustomerName ?? "Unknown"}");
            lines.Add($"Date: {order.Date ?? "Unknown"}");
            lines.Add(new string('-', 50));
            foreach (var item in order.Items ?? new List<OrderItem>())
            {
                lines.Add($"  {item.Name}: ${item.Price:F2}");
            }
            lines.Add(new string('-', 50));
            lines.Add($"Subtotal: ${order.Subtotal:F2}");
            lines.Add($"Shipping: ${order.Shipping:F2}");
            lines.Add($"Tax: ${order.Tax:F2}");
            lines.Add($"Total: ${order.Total:F2}");
            lines.Add(new string('=', 50));
            return string.Join("\n", lines);
        }
    }
}

File: src/tools/pmd-cpd/eval-repos/synthetic/go/cross_file_a.go
// Package synthetic contains Go files for CPD testing - file A for cross-file detection.
package synthetic

import (
	"fmt"
	"math"
	"strings"
)

// OrderItem represents an item in an order.
type OrderItem struct {
	Price    float64
	Quantity int
	Discount float64
	Name     string
}

// Order represents a complete order.
type Order struct {
	ID           string
	CustomerName string
	Date         string
	Items        []OrderItem
	Subtotal     float64
	Shipping     float64
	Tax          float64
	Total        float64
}

// CalculateOrderTotal calculates the total price of items.
func CalculateOrderTotal(items []OrderItem) float64 {
	var total float64
	for _, item := range items {
		price := item.Price
		quantity := item.Quantity
		if quantity == 0 {
			quantity = 1
		}
		discount := item.Discount
		itemTotal := price * float64(quantity) * (1 - discount/100)
		total += itemTotal
	}
	return math.Round(total*100) / 100
}

var shippingRates = map[string]float64{
	"US": 5.99,
	"CA": 8.99,
	"UK": 12.99,
	"DE": 14.99,
	"FR": 14.99,
	"AU": 19.99,
}

// ApplyShippingCost applies shipping cost based on country.
func ApplyShippingCost(subtotal float64, country string) float64 {
	baseRate, ok := shippingRates[country]
	if !ok {
		baseRate = 24.99
	}
	if subtotal > 100 {
		return subtotal
	}
	return subtotal + baseRate
}

var taxRates = map[string]float64{
	"CA": 0.0725,
	"NY": 0.08,
	"TX": 0.0625,
	"FL": 0.06,
	"WA": 0.065,
}

// ApplyTax applies tax based on state.
func ApplyTax(subtotal float64, state string) float64 {
	rate, ok := taxRates[state]
	if !ok {
		rate = 0.0
	}
	tax := subtotal * rate
	return math.Round((subtotal+tax)*100) / 100
}

// FormatOrderSummary formats the order summary for display.
func FormatOrderSummary(order Order) string {
	var lines []string
	lines = append(lines, strings.Repeat("=", 50))
	lines = append(lines, "ORDER SUMMARY")
	lines = append(lines, strings.Repeat("=", 50))
	id := order.ID
	if id == "" {
		id = "N/A"
	}
	customerName := order.CustomerName
	if customerName == "" {
		customerName = "Unknown"
	}
	date := order.Date
	if date == "" {
		date = "Unknown"
	}
	lines = append(lines, fmt.Sprintf("Order ID: %s", id))
	lines = append(lines, fmt.Sprintf("Customer: %s", customerName))
	lines = append(lines, fmt.Sprintf("Date: %s", date))
	lines = append(lines, strings.Repeat("-", 50))
	for _, item := range order.Items {
		lines = append(lines, fmt.Sprintf("  %s: $%.2f", item.Name, item.Price))
	}
	lines = append(lines, strings.Repeat("-", 50))
	lines = append(lines, fmt.Sprintf("Subtotal: $%.2f", order.Subtotal))
	lines = append(lines, fmt.Sprintf("Shipping: $%.2f", order.Shipping))
	lines = append(lines, fmt.Sprintf("Tax: $%.2f", order.Tax))
	lines = append(lines, fmt.Sprintf("Total: $%.2f", order.Total))
	lines = append(lines, strings.Repeat("=", 50))
	return strings.Join(lines, "\n")
}
